server/src/main/java/org/candlepin/pinsetter/tasks/HypervisorUpdateJob.java

```diff
@@ -0,0 +1,201 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.zip.DeflaterOutputStream;
+import java.util.zip.InflaterInputStream;
+
+/**
+ * Asynchronous job for refreshing the entitlement pools for specific
+ * {@link Owner}.
+ */
+public class HypervisorUpdateJob extends UniqueByOwnerJob {
+
+    private static Logger log = LoggerFactory.getLogger(HypervisorUpdateJob.class);
+    private OwnerCurator ownerCurator;
+    private ConsumerCurator consumerCurator;
+    private ConsumerResource consumerResource;
+
+    public static final String CREATE = "create";
+    public static final String DATA = "data";
+    public static final String PRINCIPAL = "principal";
+    protected static String prefix = "hypervisor_update_";
+
+    @Inject
+    public HypervisorUpdateJob(OwnerCurator ownerCurator, ConsumerCurator consumerCurator,
+            ConsumerResource consumerResource) {
+        this.ownerCurator = ownerCurator;
+        this.consumerCurator = consumerCurator;
+        this.consumerResource = consumerResource;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Executes {@link ConsumerResource#create(org.candlepin.model.Consumer, org.candlepin.auth.Principal,
+     *  java.utl.String, java.utl.String, java.utl.String)}
+     * Executes (@link ConusmerResource#updateConsumer(java.utl.String, org.candlepin.model.Consumer)}
+     * as a pinsetter job.
+     *
+     * @param context the job's execution context
+     */
+    public void toExecute(JobExecutionContext context) throws JobExecutionException {
+        try {
+            JobDataMap map = context.getMergedJobDataMap();
+            String ownerKey = map.getString(JobStatus.TARGET_ID);
+            Boolean create = map.getBoolean(CREATE);
+            Principal principal = (Principal) map.get(PRINCIPAL);
+
+            Owner owner = ownerCurator.lookupByKey(ownerKey);
+            if (owner == null) {
+                context.setResult("Nothing to do. Owner does not exist");
+                return;
+            }
+            byte[] data = (byte[]) map.get(DATA);
+            String json = decompress(data);
+            HypervisorList hypervisors = (HypervisorList) Util.fromJson(json, HypervisorList.class);
+            log.info("Hypervisor consumers for create/update: " + hypervisors.getHypervisors().size());
+
+            for (Consumer hypervisor : hypervisors.getHypervisors()) {
+                if (hypervisor.getHypervisorId() != null &&
+                        hypervisor.getHypervisorId().getHypervisorId() != null) {
+                    String hypervisorId = hypervisor.getHypervisorId().getHypervisorId();
+                    hypervisor.setType(new ConsumerType(ConsumerTypeEnum.HYPERVISOR));
+                    Consumer knownHost = consumerCurator.getHypervisor(hypervisorId, owner);
```

Comment:
Big performance issues here. We recently worked to fix this in the old API call as the long transactions causes big deadlock and timeout issues. You can see this in the old API call where we lookup hypervisorConsumersMap, guestConsumersMap, and guestHypervisorConsumers. 

You'll see guestHypervisorConsumers eventually passed to consumerResource.performConsumerUpdates(). I did not add it to update consumer resource method but it looks like we might need to here. Not passing this through is likely resulting in one lookup per guest ID as well, so thousands and thousands of DB queries whereas the old API was doing ~4.

I would be very worthwhile to generate a JSON file or something for use with virt-whom and test this with about 2000 hypervisors each with 10-15 guest IDs, just as a reasonable real world scenario. If we want to plan for future I'd bump that up a lot. I'd suggest doing this before making the performance changes so you can see them taking effect.